### ISSUE TITLE
Segmentation fault after 3.1.27 update

### ISSUE THREAD

**Issue by ceifa** [emscripten-core/emscripten#18401]
I'm receiving a segfault since 3.1.27. Older versions works great. https://github.com/ceifa/wasmoon/actions/runs/3724920493/jobs/6317398427

**Version of emscripten/emsdk:**

```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.27 (afa75f342eef4d925172479afa8e6233eb0ae5a9)
clang version 16.0.0 (https://github.com/llvm/llvm-project dee009d3b5c71a340eb4ebb3cfb49f6015a789cc)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/c31/emsdk/upstream/bin
```

**How I'm building the wasm:**
```
emcc \
    -s WASM=1 -o ./build/glue.js \
    -s MODULARIZE=1 \
    -s ALLOW_TABLE_GROWTH=1 \
    -s EXPORT_NAME="initWasmModule" \
    -s ALLOW_MEMORY_GROWTH=1 \
    -s STRICT=1 \
    -s EXPORT_ES6=1 \
    -s NODEJS_CATCH_EXIT=0 \
    -s NODEJS_CATCH_REJECTION=0 \
    -s MALLOC=emmalloc
```

**Comment by sbc100** [comment]
Could well be the new default stack size.   See the Changelog.  Can you try building with `-sSTACK_SIZE=5mb`?

**Comment by ceifa** [comment]
Tried it and got the same error!

```
emcc \
    -s WASM=1 -o ./build/glue.js \
    -s MODULARIZE=1 \
    -s ALLOW_TABLE_GROWTH=1 \
    -s EXPORT_NAME="initWasmModule" \
    -s ALLOW_MEMORY_GROWTH=1 \
    -s STRICT=1 \
    -s EXPORT_ES6=1 \
    -s NODEJS_CATCH_EXIT=0 \
    -s NODEJS_CATCH_REJECTION=0 \
    -s MALLOC=emmalloc \
   -sSTACK_SIZE=5mb
```

**Comment by sbc100** [comment]
In that case maybe you can bisecting so we can see what might have caused the regression: https://emscripten.org/docs/contributing/developers_guide.html#bisecting

**Comment by gottfriedleibniz** [comment]
Referencing: https://github.com/emscripten-core/emscripten/commit/ce4c405700d96a0b945f4937e414fd7c114fb2a4: Updating the build script to include: `-s USE_ES6_IMPORT_META=1 -s ENVIRONMENT=node` fixes the action. Increasing `STACK_SIZE` also would not hurt given the nature of the project.

Actual actions to take w/ this knowledge, if any, is beyond my pay grade.

**Comment by ceifa** [comment]
@gottfriedleibniz you are right, changing the environment to node fix the error on the action. The problem is that the project should emit code for all environments, so this is not the right solution to my problem.

I reviewed the https://github.com/emscripten-core/emscripten/commit/ce4c405700d96a0b945f4937e414fd7c114fb2a4 commit and it doesn't have anything that changes the wasm behavior, maybe @kleisauke can help us understand?

**Comment by kleisauke** [comment]
Thanks for the ping. That commit has caused a regression on Safari (#18357 - https://bugs.webkit.org/show_bug.cgi?id=223533), but I'm not aware that it caused a (suspicious) segfault on Node.js.

Since this was originally triaged as a Jest issue (https://github.com/ceifa/wasmoon/issues/65), does this segfault only occur when running with Jest? FWIW, I didn't notice this issue with [Mocha](https://mochajs.org/), both on the browser and on Node.js.

> -s ENVIRONMENT=node

Given that this was mentioned as a workaround, I suspect it could be related to the `async` function that is emitted for multi-environment builds.
https://github.com/emscripten-core/emscripten/blob/bfabc4e0b82539ba46dfdd1675198dab32a25cfb/emcc.py#L3757-L3762

**Comment by ceifa** [comment]
> Since this was originally triaged as a Jest issue ([ceifa/wasmoon#65](https://github.com/ceifa/wasmoon/issues/65)), does this segfault only occur when running with Jest? FWIW, I didn't notice this issue with [Mocha](https://mochajs.org/), both on the browser and on Node.js.

Yes, I could only reproduce it with jest.

**Comment by ceifa** [comment]
> Since this was originally triaged as a Jest issue ([ceifa/wasmoon#65](https://github.com/ceifa/wasmoon/issues/65)), does this segfault only occur when running with Jest? FWIW, I didn't notice this issue with [Mocha](https://mochajs.org/), both on the browser and on Node.js.

Yes, I could only reproduce it with jest. Can it be a jest problem?
